app/services/atmosphere/billing_service.rb

```diff
@@ -72,48 +84,45 @@ def self.bill_appliance(appl, billing_time, message, apply_prepayment = true)
-    def self.calculate_charge_for_appliance(appl, billing_time, apply_prepayment = true)
-      # By default, the billing process should extend the validity of the appliance until now+1.hour.
-      billable_time = self.calculate_billable_time(appl, billing_time, apply_prepayment)
+    # - This is usually true, but will be false when calculating final charge for a deployment which is being shut down.
+    def self.calculate_charge_for_deployment(dep, billing_time, apply_prepayment = true)
+      # By default, the billing process should extend the validity of the deployment until now+1.hour.
+      billable_time = self.calculate_billable_time(dep, billing_time, apply_prepayment)
 
-      amount_due = 0
+      vm = dep.virtual_machine # Shorthand
+      # Find out how many appliances are using this VM and split costs equally
+      os_family = dep.appliance.appliance_type.os_family
+      hourly_charge = (vm.virtual_machine_flavor.get_hourly_cost_for(os_family)/vm.appliances.count).round #TODO: figure out if we should limit this to appliances with billing_state == :prepaid
+      Rails.logger.debug("Calculated hourly charge for using VM #{vm.id} is #{hourly_charge}. This VM currently has #{vm.appliances.count} appliances using it.")
+      charge = (hourly_charge*billable_time).round
 
-      # Iterate over appliance vms (there can be many, if the appliance is shared)
-      appl.virtual_machines.each do |vm|
-        # Find out how many appliances are using this VM and split costs equally
-        hourly_charge = (vm.virtual_machine_flavor.hourly_cost/vm.appliances.count).round #TODO: figure out if we should limit this to appliances with billing_state == :prepaid
-        Rails.logger.debug("Calculated hourly charge for using VM #{vm.id} is #{hourly_charge}. This VM currently has #{vm.appliances.count} appliances using it.")
-        charge = (hourly_charge*billable_time).round
-        amount_due += charge
-      end
-      # Return amount due
-      amount_due
+      # Return anticipated charge
+      charge
     end
 
-    # This scans for :expired appliances and figures out what to do with the underlying VMs, following fund policies
+    # This scans for :expired deployments and figures out what to do with the underlying VMs, following fund policies
     def self.apply_funding_policy
       VirtualMachine.where(managed_by_atmosphere: true).all.each do |vm|
-        # If this VM has at least one prepaid appliance, it must not be touched. For safety's sake, we will also not touch appliances whose billing state is flagged as erroneous
-        if vm.appliances.select {|appl| ['prepaid', 'error'].include? appl.billing_state}.count > 0
+        # If this VM has at least one prepaid deployment, it must not be touched. For safety's sake, we will also not touch deployments whose billing state is flagged as erroneous
+        if vm.deployments.select {|dep| ['prepaid', 'error'].include? dep.billing_state}.count > 0
```

Comment:
Hmm, it's a subjective opinion. I find it easier to analyze code when I don't have to hop around the tree trying to figure out what each method does. I agree that very complex conditions merit externalization but I don't think we've crossed this line here. :)